src/Nethermind/Nethermind.Core/Extensions/SpanExtensions.cs

```diff
@@ -25,93 +29,89 @@ public static string ToHexString(this in Span<byte> span, bool withZeroX, bool n
         }
 
         [DebuggerStepThrough]
-        private static string ToHexViaLookup(in Span<byte> span, bool withZeroX, bool skipLeadingZeros, bool withEip55Checksum)
+        private static string ToHexViaLookup(Span<byte> bytes, bool withZeroX, bool skipLeadingZeros, bool withEip55Checksum)
         {
-            int leadingZeros = skipLeadingZeros ? CountLeadingZeros(span) : 0;
-            char[] result = new char[span.Length * 2 + (withZeroX ? 2 : 0) - leadingZeros];
-            string? hashHex = null;
             if (withEip55Checksum)
             {
-                hashHex = Keccak.Compute(span.ToHexString(false)).ToString(false);
+                return ToHexStringWithEip55Checksum(bytes, withZeroX, skipLeadingZeros);
             }
 
+            int leadingZeros = skipLeadingZeros ? Bytes.CountLeadingZeros(bytes) : 0;
+            int length = bytes.Length * 2 + (withZeroX ? 2 : 0) - leadingZeros;
+
+            char[] charArray = ArrayPool<char>.Shared.Rent(length);
+
+            ref byte input = ref Unsafe.Add(ref MemoryMarshal.GetReference(bytes), leadingZeros / 2);
+            Bytes.OutputBytesToCharHex(ref input, bytes.Length, ref charArray[0], withZeroX, leadingZeros);
+
+            string result = new string(charArray.AsSpan(0, length));
+            ArrayPool<char>.Shared.Return(charArray);
+
+            return result;
+        }
+
+        private static string ToHexStringWithEip55Checksum(Span<byte> bytes, bool withZeroX, bool skipLeadingZeros)
+        {
+            string hashHex = Keccak.Compute(bytes.ToHexString(false)).ToString(false);
+
+            int leadingZeros = skipLeadingZeros ? Bytes.CountLeadingZeros(bytes) : 0;
+            int length = bytes.Length * 2 + (withZeroX ? 2 : 0) - leadingZeros;
+            if (leadingZeros >= 2)
+            {
+                bytes = bytes[(leadingZeros / 2)..];
+            }
+            char[] charArray = ArrayPool<char>.Shared.Rent(length);
+
+            Span<char> chars = charArray.AsSpan(0, length);
+
             if (withZeroX)
             {
-                result[0] = '0';
-                result[1] = 'x';
+                // In reverse, bounds check on [1] will elide bounds check on [0]
+                chars[1] = 'x';
```

Comment:
QQ: I'm guessing, there is some kind of optimization from setting it in reverse?

Reply:
Nevermind...